Add C4 Streaming dataset

abhi-mosaic opened the pull request:
This PR adds an adapter for HF streaming datasets, so that we can train with DDP and sharding. There is also a small QoL fix for DeepSpeed.

I've also added a new GPT3-125m YAML which can be the start of our new GPT3 family of benchmarks. I think we should deprecate the GPT2 collection of YAMLs (-52m, -83m -125m) in the near future, say Composer v0.5.

Some TODOs, which can be follow-up PRs:

* Add the rest of the GPT3 family YAMLs
* Update the GPT2 model with QoL fixes
* Update the BERT-base benchmark to use C4
* Deprecate / combine the old `LMDataset` object

hanlint commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,246 @@
+        else:
+            raise ValueError(f"Unknown group_method: '{self.group_method}'")
+
+    def initialize_object(self, batch_size: int, dataloader_hparams: DataloaderHparams) -> DataSpec:
```
is there a way we can move all of this logic into a separate class? 

the motivation being that external users not using YAHP should be able to create their StreamingDataset and pass it to our trainer directly?

abhi-mosaic replied:
Yes, I could definitely move some of the first few steps like sharding, and the overwritten __len__ into a class, but if users want to build their own custom dataloader, they would still have to build the collate functions and pass those correctly to their dataloader right? I'm also wondering if the preprocessing steps like (shuffle, tokenize, group) would belong in the class, or if again it is user-specific. I guess I'm skeptical how much I can abstract away, haha.

Looking at Vision, I think there are classes like `ImageFolder(..) -> Dataset` that you can start from, but building a dataloader is still pretty task-specific.

For users that want to build their own streaming dataloaders, I think the upcoming Vision Webdatasets PR would be the better example to follow, and that PR comes with helper classes too. 

WDYT? I'm happy to take a pass at making a light HFStreamingDataset class, or punt it to a future PR?

hanlint replied:
Update: we discussed offline, and will refactor a standalone dataset class for say notebook or script users.

abhi-mosaic commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
+    if isinstance(batch, dict):
+        chunked = {k: v.chunk(n_microbatches) for k, v in batch.items()}
+        num_chunks = len(list(chunked.values())[0])
```
Apparently `torch.chunk` is not guaranteed to return the number, if for some reason we request more chunks than the length of the list. I'm not sure if we have any asserts anywhere for this... but I can throw a ValueError if that makes sense?

abhi-mosaic commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+    config_name: Optional[str] = hp.optional(
+        "If required, the specific configuration of the dataset that you would like to use.", default=None)
+    split: str = hp.optional("What split of the dataset to use (e.g. 'train' or 'validation' or 'test')", default=None)
+    max_shards: int = hp.optional("Max number of shards, used to deterministically reduce dataset size.", default=-1)
```
I think `max_shards` can be deleted as we're converting this to a C4-specific dataset class.

abhi-mosaic commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+    shuffle_buffer_size: int = hp.optional("How many incoming data samples should be kept in a buffer to shuffle.",
+                                           default=10000)
+    seed: int = hp.optional("Which seed to use to generate train and validation splits.", default=5)
+    shuffle: bool = hp.optional("Whether to shuffle the dataset for each epoch.", default=True)
```
Yep! Will fix.

abhi-mosaic commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+        "Max number of post-processed samples, note that the subset will depend on seed and world size.", default=-1)
+    tokenizer_name: str = hp.optional("The name of the tokenizer to preprocess text with.", default=None)
+    max_seq_len: int = hp.optional("The max sequence length of each token sample.", default=None)
+    group_method: str = hp.optional("How to group text samples into token samples.", default=None)
```
Yes, will add a list of choices. The choices are not HF-related, it's just two modes that we have defined for ourselves to determine how to treat the incoming stream of text samples (merge samples to produce full max_sequence_length samples, or keep samples separate and pad)

abhi-mosaic commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+                n_shards = self.max_shards if self.max_shards > 0 else n_shards
+                return n_shards * samples_per_shard // dist.get_world_size()
+        except:
+            raise NotImplementedError
```
Will refactor this so it is C4 specific.

ravi-mosaicml commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+    """Defines a generic dataset class for autoregressive and masked language models."""
+
+    name: str = hp.optional("Name of the dataset to load.", default=None)
+    config_name: Optional[str] = hp.optional(
```
Can you give examples or link to HF docs for what a configuration name would be?

abhi-mosaic replied:
Removing configuration name for now.

ravi-mosaicml commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+    config_name: Optional[str] = hp.optional(
+        "If required, the specific configuration of the dataset that you would like to use.", default=None)
+    split: str = hp.optional("What split of the dataset to use (e.g. 'train' or 'validation' or 'test')", default=None)
+    max_shards: int = hp.optional("Max number of shards, used to deterministically reduce dataset size.", default=-1)
```
What would -1 mean?

abhi-mosaic replied:
Removing `max_shards` as we are making this example C4 specific.

ravi-mosaicml commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+                                      default=None)
+    mlm_probability: float = hp.optional("If using masked language modeling, the probability to mask tokens with.",
+                                         default=0.15)
+    shuffle_buffer_size: int = hp.optional("How many incoming data samples should be kept in a buffer to shuffle.",
```
Can you add in saying this parameter has an effect only if shuffle is True?

ravi-mosaicml commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+    drop_last: bool = hp.optional("Whether to drop the last samples for the last batch.", default=False)
+
+    def validate(self):
+        assert self.name is not None, "Must provide a dataset name."
```
For validation, can you using ValueErrors? assert should not be used for data validation (see https://github.com/mosaicml/composer/blob/dev/STYLE_GUIDE.md#4-use-of-assert)

ravi-mosaicml commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+            import transformers
+        except ImportError:
+            raise ImportError('HuggingFace transformers and datasets are not installed. '
+                              'Please install with `pip install mosaicml-composer[nlp]`')
```
```suggestion
                              'Please install with `pip install composer[nlp]`')
```

ravi-mosaicml commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+        except ImportError:
+            raise ImportError('HuggingFace transformers and datasets are not installed. '
+                              'Please install with `pip install mosaicml-composer[nlp]`')
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(self.tokenizer_name)  #type: ignore (thirdparty)
```
What is the type error?

abhi-mosaic replied:
Code leftover from previous implementation, removing the ignore.

ravi-mosaicml commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+            raise ImportError('HuggingFace transformers and datasets are not installed. '
+                              'Please install with `pip install mosaicml-composer[nlp]`')
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(self.tokenizer_name)  #type: ignore (thirdparty)
+        self.config = transformers.AutoConfig.from_pretrained(self.tokenizer_name)  #type: ignore (thirdparty)
```
Same

ravi-mosaicml commented on composer/datasets/hf_streaming.py:
```diff
@@ -0,0 +1,243 @@
+                                                                      mlm_probability=self.mlm_probability)
+        # Return DataSpec
+        return DataSpec(dataloader=dataloader_hparams.initialize_object(
+            dataset=sized_iterable_dataset,
```
Can you pass num_samples and num_tokens into the DataSpec?

abhi-mosaic replied:
I'm not sure if we are quite ready yet... there's still some discussion with NLP team as to what constitutes a sample / a token when there is padding and concatenation involved. I can bring it up with relevant team members and make a follow up PR.

ravi-mosaicml commented on composer/trainer/trainer_hparams.py:
```diff
@@ -277,38 +277,39 @@ class TrainerHparams(hp.Hparams):
 
         if self.deepspeed is not None:
-            zero_stage = cast(int, self.deepspeed.get("zero_stage", 0))
+            self.deepspeed["zero_stage"] = cast(int, self.deepspeed.get("zero_stage", 0))
```
Are these casts necessary?

abhi-mosaic replied:
Probably not, carried over from previous code. Will remove and see what happens.

abhi-mosaic replied:
Actually turns out they are... because the `self.deepspeed` config is of type `[str, JSON]`, and so the values we want to default have to be cast to int before the validation that happens below.

abhi-mosaic commented:
This PR has been refactored, so instead of a general HF streaming adapter (which is a bit unfeasible, given that HF datasets are user-generated and do not have a consistent implementation of shards or sample dicts), this PR now adds a specific `C4Dataset` that happens to be backed by HF Datasets.

The intent is for `C4Dataset` to be an example for users who want to train with their own HF datasets, but they will have to construct their `CustomDataset` class independently. This is similar to how our `ImageNet` dataset is an example of a vision dataset that happens to be backed by `torchvision.datasets.ImageFolder`.

In the future, we can try to abstract some things away, and/or coordinate with HF upstream to build better sharding support, but right now `C4Dataset` is just a standalone example.

hanlint commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+        self.shuffle_buffer_size = shuffle_buffer_size
+        self.seed = seed
+
+        # Metadata
```
just to check, this is specific for C4 correct?

can you encode it in a single `dict`? e.g.
```
c4_metadata = {"train": {"num_shards" = 1024, .. }, "validation": {"num_shards": 8, ..}}
```

abhi-mosaic replied:
Yes, this is very specific to the C4 dataset built/uploaded by AllenAI and hosted on HuggingFace. I can simplify it to a dict.

hanlint commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+        self.rank = dist.get_global_rank()
+        self.max_samples_per_device = self.max_samples // self.world_size
+        new_max_samples = self.max_samples_per_device * self.world_size
+        if new_max_samples != self.max_samples:
```
might be more readable to change this to test for divisibility directly:
```
if self.max_samples % self.world_size != 0:
     warning!
     self.max_samples = self.max_samples_per_device * self.world_size
```

hanlint commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+            max_length = None
+        return self.tokenizer(text_batch["text"], truncation=truncation, padding=padding, max_length=max_length)
+
+    def _group_tokens(self, token_batch):
```
as some docs for what this is actually doing? input/outputs etc?

hanlint commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+            text_batch[k] = v[device_offset:device_offset + 1]
+        return text_batch
+
+    def _shard_dataset(self, dataset):
```
even those these are private functions, i think we should have good docstrings because:
1) any step of the data preprocessing is important for users to understand
2) users may want to extend this to other datasets themselves

hanlint commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+            if self.world_size % self.num_shards != 0:
+                raise ValueError(f"Cannot evenly split {self.num_shards} shards among {self.world_size} devices")
+            devices_per_shard = self.world_size // self.num_shards
+        elif self.num_shards % self.world_size != 0:
```
since `num_shards` is defined within the functioni (e.g. `1024`), does that mean that using the streaming dataset with 3 or 6 or 7 GPUs is not supported?

abhi-mosaic replied:
Sadly, yes... in general it gets pretty tricky to handle non-divisible number of shards, I can think of two options:
* truncate / duplicate some shards to make `new_num_shards` divisible
* figure out some way to iterate through one set of shards normally, and then a second set of shards (the leftovers) via subsampling... but this seems a bit complicated.

For context, the `webdataset` package also just splits the shard list among all CPU workers and assumes that the `num_shards` is divisible.

I think these limitations can be overcome, but I wonder if this is the best place for it, or if we can follow up with some more general adapter that can correct both HF and Webdataset IterableDatasets.

hanlint replied:
Ok with not addressing, just wanted clarification.

hanlint reviewed (commented):
Did mostly a code style/ readability review. Will have to rely on more knowledgable (e.g. @moinnadeem ) to check the functional correctness.

moinnadeem commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+    mlm: bool = hp.optional("Whether or not to use masked language modeling.", default=False)
+    mlm_probability: float = hp.optional("If `mlm=True`, the probability that tokens are masked.", default=0.15)
+    shuffle: bool = hp.optional(
+        "Whether to shuffle the samples in the dataset. Currently, shards are assigned deterministically to devices, but shuffling affects the order of shards (per-device), and the order of samples via (per-device) shuffle buffers.",
```
Why would shuffling affect the order of the shards, rather than just shuffle within a shard?

abhi-mosaic replied:
Sadly this is what the HF IterableDatset.shuffle function does, and there's no flag to turn off shard shuffling: https://huggingface.co/docs/datasets/package_reference/main_classes.html#datasets.IterableDataset.shuffle. 

However I could potentially move the shuffle op to the very end of the dataset definition, after `.take(samples_per_device)`. This would hypothetically stop the shard shufffling and mean that `max_samples` will always refer to the same subset regardless of seed. I think this is preferable...

... but note that `max_samples` is still affected by the number of devices! And that is impossible to fix... because if using 1 device the first 10 samples are all from shard1, but if using 2 devices the first 10 samples are (5 from shard1, 5 from shard2).

moinnadeem commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+
+        # Build tokenizer
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(tokenizer_name)
+        if self.tokenizer_name in ["gpt2"]:
```
Can we just check `if self.tokenizer.pad_token_id is None` rather than checking if the tokenizer name matches a very specific tokenizer?

abhi-mosaic replied:
Yup! As long as you think that the EOS -> PAD replacement is generally safe, and not just a GPT2 thing.

moinnadeem commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+        # Map text samples to token samples
+        # NOTE: Mapping is executed in batched mode for better CPU utilization,
+        # but the returned dataset is still an iterable over tokenized samples
+        text_sample_batch_size = 1000
```
Can we make this an hparam of the dataset?

abhi-mosaic replied:
I'm a little wary of exposing this as it's kind of a CPU utilization thing and seems like clutter, I think it's also the HF default (so maybe I can even remove it from my code). What do you think?

abhi-mosaic replied:
Closing for now, I think it will be simpler to keep this hparam internal as it's a perf thing.

moinnadeem commented on composer/datasets/c4.py:
```diff
@@ -0,0 +1,333 @@
+            # but the returned dataset is still an iterable over tokenized samples.
+            # NOTE: Depending on the 'group_method', this step may alter the number of
+            # token samples in the dataset, and may mix neighboring token samples together.
+            token_sample_batch_size = 1000
```
Can we make this an hparam of the dataset?

moinnadeem reviewed (commented):
Added a few comments, correctness looks good to me from what I can tell! Good work Abhi!

abhi-mosaic commented:
Hey @hanlint @moinnadeem , sorry for the delay on this PR. I think I should have addressed all the comments. Hope we can get this merged today! And then I can work with @moinnadeem to update the BERT yaml to use C4, as well as convert some of the large GPT2 yamls (> 125M) to GPT3.

hanlint reviewed (approved):
LGTM from code quality stand point, defer to @moinnadeem for NLP correctness

abhi-mosaic commented:
Also, I'm seeing a lot of new pyright errors.. but I don't think there are any related to `c4.py`. Please let me know if this is expected / what I can do.

abhi-mosaic commented:
Turns out the self documenting f-string syntax `f"{var=}` did not appear until python3.8, so our python3.7 tests were failing 🙃 . Reverted...
